wagtail/wagtailcore/models.py

```diff
@@ -107,21 +107,59 @@ def find_for_request(request):
         NB this means that high-numbered ports on an extant hostname may
         still be routed to a different hostname which is set as the default
         """
+
+        # default site q
+        q = Q(is_default_site=True)
+
+        # hostname q
         try:
-            hostname = request.META['HTTP_HOST'].split(':')[0]  # KeyError here goes to the final except clause
-            try:
-                # find a Site matching this specific hostname
-                return Site.objects.get(hostname=hostname)  # Site.DoesNotExist here goes to the final except clause
-            except Site.MultipleObjectsReturned:
-                # as there were more than one, try matching by port too
-                port = request.META['SERVER_PORT']  # KeyError here goes to the final except clause
-                return Site.objects.get(hostname=hostname, port=int(port))
-                # Site.DoesNotExist here goes to the final except clause
-        except (Site.DoesNotExist, KeyError):
-            # If no matching site exists, or request does not specify an HTTP_HOST (which
-            # will often be the case for the Django test client), look for a catch-all Site.
-            # If that fails, let the Site.DoesNotExist propagate back to the caller
-            return Site.objects.get(is_default_site=True)
+            hostname = request.META['HTTP_HOST'].split(':')[0]
+            q |= Q(hostname=hostname)
+        except KeyError:
+            pass
+
+        # get all matching sites
+        # order by is_default_site so if there is a default it is either first or last
+        site_list = list(Site.objects.filter(q).order_by('is_default_site'))
+
+        # only one result (fastpath)
+        if len(site_list) == 1:
+            return site_list[0]
+
+        # DoesNotExist
+        if len(site_list) == 0:
+            raise Site.DoesNotExist()
+
+        # find the default_site (if any)
+        # remove from list if hostname does not match
+        # we need to check both ends as different dbs store and order booleans differently (see django #19726)
+        default_site = None
```

Comment:
We would still need to find which is the default site, and that
would require another O(n) search. Then removing it alone with a whole list
comprehension would be another O(n) (there will only ever be one Site with
a non-matching hostname, and there could be none!).

The way I have written it is O(1): no matter how big the site_list, the
default site will be either first or last (or neither).

I agree that it is more verbose this way, but I fail to see negligable
impact. The catalyst for my creating this PR was hitting the db twice for
every non-default site page view, esentially doubling the time it takes to
work out which site the request is for.

Maybe you could provide a full alternate solution including handling the
default site id and removal? What you have posted would not produce the
same results as the PR code. But I am interested to see what you mean fully.

On Friday, 25 December 2015, Karl Hobley <notifications@github.com> wrote:

> In wagtail/wagtailcore/models.py
> <https://github.com/torchbox/wagtail/pull/1963#discussion_r48418370>:
>
> > +        # get all matching sites
> > +        # order by is_default_site so if there is a default it is either first or last
> > +        site_list = list(Site.objects.filter(q).order_by('is_default_site'))
> > +
> > +        # only one result (fastpath)
> > +        if len(site_list) == 1:
> > +            return site_list[0]
> > +
> > +        # DoesNotExist
> > +        if len(site_list) == 0:
> > +            raise Site.DoesNotExist()
> > +
> > +        # find the default_site (if any)
> > +        # remove from list if hostname does not match
> > +        # we need to check both ends as different dbs store and order booleans differently (see django #19726)
> > +        default_site = None
>
> I think we should simplify this a bit. Perhaps a list comprehension?
>
> site_list = [site for site in site_list if site.hostname == hostname]
> ``
>
> Negligable performance impact and much more expressive.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/torchbox/wagtail/pull/1963/files#r48418370>.
>